src/main/java/io/cdap/plugin/successfactors/source/transport/SuccessFactorsTransporter.java

```diff
@@ -0,0 +1,232 @@
+  private final String username;
+  private final String password;
+
+  public SuccessFactorsTransporter(String username, String password) {
+    this.username = username;
+    this.password = password;
+
+  }
+
+  /**
+   * Calls the Successfactors entity for the given URL and returns the respective response.
+   * Supported calls are:
+   * - testing the URL correctness
+   * - fetching the SuccessFactors entity metadata
+   * - fetching the total available records count
+   *
+   * @param endpoint  type of URL
+   * @param mediaType mediaType for Accept header property, supported types are 'application/json' & 'application/xml'
+   * @param fetchType type of call i.e. TEST / METADATA / COUNT, used for logging purpose.
+   * @return {@code SuccessFactorsResponseContainer}
+   * @throws TransportException any http client exceptions are wrapped under it
+   */
+  public SuccessFactorsResponseContainer callSuccessFactorsEntity(URL endpoint, String mediaType, String fetchType)
+    throws TransportException {
+
+    try {
+      LOG.debug(ResourceConstants.DEBUG_CALL_SERVICE_START.getMsgForKey(fetchType));
+      Response res = transport(endpoint, mediaType);
+      LOG.debug(ResourceConstants.DEBUG_CALL_SERVICE_END.getMsgForKey(fetchType));
+
+      return prepareResponseContainer(res);
+    } catch (IOException ioe) {
+      throw new TransportException(ResourceConstants.ERR_CALL_SERVICE_FAILURE.getMsgForKey(), ioe);
+    }
+  }
+
+  /**
+   * Calls the Successfactors entity to fetch the records with subsequent retries in case of failure.
+   * Retry modes are:
+   * - any HTTP code equal or above 500
+   * - max retry is 3 times
+   *
+   * @param endpoint record fetch URL
+   * @return {@code SuccessFactorsResponseContainer}
+   * @throws TransportException   any http client exceptions are wrapped under it
+   * @throws InterruptedException any error in case of retry suspension state
+   */
+  public SuccessFactorsResponseContainer callSuccessFactorsWithRetry(URL endpoint)
+    throws TransportException, InterruptedException, IOException {
+
+    LOG.debug(ResourceConstants.DEBUG_CALL_SERVICE_START.getMsgForKey(SuccessFactorsService.DATA));
+    Response res = retrySapTransportCall(endpoint, MediaType.APPLICATION_JSON);
+    LOG.debug(ResourceConstants.DEBUG_CALL_SERVICE_END.getMsgForKey(SuccessFactorsService.DATA));
+
+    try {
+      return prepareResponseContainer(res);
+    } catch (IOException ioe) {
+      res.close();
+      throw new TransportException(ResourceConstants.ERR_CALL_SERVICE_FAILURE.getMsgForKey(), ioe);
+    }
+  }
+
+  /**
+   * Calls the given URL with retry logic.
+   *
+   * @param endpoint  record fetch URL
+   * @param mediaType mediaType for Accept header property
+   * @return {@code Response}
+   * @throws IOException if all retries fail
+   */
+  public Response retrySapTransportCall(URL endpoint, String mediaType) throws IOException {
+    final Response[] res = new Response[1];
```

Comment:
why is res an array?

Reply:
Having a local variable used in lambda expression should be final or effectively final and I couldn't assign a value to the final variable inside the lambda expression, so preferred to use an array.

Have now replaced the array with a global object for `response`. 288b022add0e0a7a19652bd3d4efff4b18f3bf2f